src/librustc_mir/interpret/eval_context.rs

```diff
@@ -1314,47 +1316,117 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> EvalContext<'a, 'mir, 'tcx, M
+    ) -> EvalResult<'tcx> {
+        trace!("validate scalar: {:#?}, {:#?}, {:#?}", value, size, scalar);
+        use std::ops::RangeBounds;
+        use std::fmt::Debug;
+        fn in_range<'tcx>(value: Scalar, size: Size, bound: impl RangeBounds<u128> + Debug) -> EvalResult<'tcx, bool> {
+            match value {
+                Scalar::Bits {..} => {
+                    let bits = value.to_bits(size)?;
+                    Ok(bound.contains(&bits))
+                },
+                Scalar::Ptr(_) => {
+                    // we cannot reason about Ptr numeric ranges other than zero
+                    if bound.start_bound() == Bound::Included(&0) && bound.end_bound() == Bound::Included(&0) {
+                        Ok(false)
+                    } else {
+                        Ok(true)
+                    }
+                }
+            }
+        }
+        let (lo, hi) = scalar.valid_range.clone().into_inner();
+        if lo > hi {
+            if in_range(value, size, ..=hi)? || in_range(value, size, lo..)? {
+                Ok(())
+            } else {
+                err!(ValidationFailure(format!(
+                    "encountered {:?} but expected something in the range {:?} or {:?}",
+                    value,
+                    ..=hi,
+                    lo..,
+                )))
+            }
+        } else {
+            if in_range(value, size, scalar.valid_range.clone())? {
+                Ok(())
+            } else {
+                err!(ValidationFailure(format!(
+                    "encountered {:?} but expected something in the range {:?}",
+                    value,
+                    scalar.valid_range,
+                )))
+            }
+        }
+    }
+
     pub fn validate_ptr_target(
         &self,
         ptr: Pointer,
         ptr_align: Align,
-        ty: Ty<'tcx>
+        mut layout: TyLayout<'tcx>,
     ) -> EvalResult<'tcx> {
-        match ty.sty {
-            ty::TyBool => {
-                self.memory.read_scalar(ptr, ptr_align, Size::from_bytes(1))?.to_bool()?;
-            }
-            ty::TyChar => {
-                let c = self.memory.read_scalar(ptr, ptr_align, Size::from_bytes(4))?.to_bits(Size::from_bytes(4))? as u32;
-                match ::std::char::from_u32(c) {
-                    Some(..) => (),
-                    None => return err!(InvalidChar(c as u128)),
+        self.memory.dump_alloc(ptr.alloc_id);
+        trace!("validate_ptr_target: {:#?}", layout);
+        if let layout::Variants::Tagged { .. } = layout.variants {
+            let variant_index = self.read_discriminant_as_variant_index(
+                Place::from_ptr(ptr, ptr_align),
+                layout,
+            )?;
+            layout = layout.for_variant(self, variant_index);
+            trace!("variant layout: {:#?}", layout);
+        }
+        match layout.fields {
+            // primitives are unions with zero fields, because that's totally obvious and documented
```

Comment:
might be worth pointing at documentation/reference? unless it's worth refactoring this weirdness out?

Reply:
Maybe it should check `layout.abi` before looking at `layout.fields`?

Reply:
I documented it ;)

Reply:
To me this sounds like indeed, it should first check `layout.abi`?